src/menelaus/concept_drift/md3.py

```diff
@@ -0,0 +1,358 @@
+        self.curr_margin_density = (self.forgetting_factor * self.curr_margin_density + 
+                                    (1 - self.forgetting_factor) * margin_inclusion_signal)
+        
+        warning_level = np.abs(self.curr_margin_density - self.reference_distribution["md"])
+        warning_threshold = self.sensitivity * self.reference_distribution["md_std"]
+        print("warning level:", warning_level)
+        print("warning threshold:", warning_threshold)
+        
+        if warning_level > warning_threshold:
+            self.drift_state = "warning"
+            self.waiting_for_oracle = True
+        
+    def give_oracle_label(self, labeled_sample):
+        """
+        Provide the detector with a labeled sample to confirm or rule out drift. Once a
+        certain number of samples is accumulated, drift can be confirmed or ruled out. If
+        drift is confirmed, retraining will be initiated using these samples, and the 
+        reference distribution will be updated accordingly.
+
+        Args:
+            labeled_sample (DataFrame): labeled data sample
+        """
+        
+        if self.waiting_for_oracle != True:
+            raise ValueError(
+                """give_oracle_labels method can be called only when a drift warning has
+                been issued and drift needs to be confirmed or ruled out."""
+            )
+            
+        if len(labeled_sample) != 1:
+            raise ValueError(
+                """This method is only available for data inputs in the form of 
+                a Pandas DataFrame with exactly 1 record."""
+            )
+        
+        labeled_columns = list(labeled_sample.columns)
+        feature_columns = list(self.reference_batch_features.columns)
+        target_column = list(self.reference_batch_target.columns)
+        reference_columns = feature_columns + target_column
+        if len(labeled_columns) != len(reference_columns) or set(labeled_columns) != set(reference_columns):
+            raise ValueError(
+                """give_oracle_labels method can be called only with a sample containing
+                the same number and names of columns as the original reference distribution."""
+            )
+            
+        self.drift_state = None
+            
+        if self.oracle_data is None:
+            self.oracle_data = labeled_sample
+        else:
+            self.oracle_data = pd.concat([self.oracle_data, labeled_sample], ignore_index = True)
+            
+        if len(self.oracle_data) == self.oracle_data_length_required:
+            X_test, y_test = self.oracle_data[feature_columns], self.oracle_data[target_column]
+            y_pred = self.classifier.predict(X_test)
+            acc_labeled_samples = accuracy_score(y_test, y_pred)
+            
+            # TODO: do absolute value here or no? algo in paper does not use absolute value
+            drift_level = np.abs(self.reference_distribution["acc"] - acc_labeled_samples)
```

Comment:
```suggestion
            drift_level = self.reference_distribution["acc"] - acc_labeled_samples #TODO: absolute value?
```
Since we had discussed sticking with the paper's "opinion" and potentially modifying later.